src/burnchains/affirmation.rs

```diff
@@ -0,0 +1,3315 @@
+    burnchain_db: &mut BurnchainDB,
+    indexer: &B,
+    reward_cycle: u64,
+    burnchain: &Burnchain,
+) -> Result<(), Error> {
+    debug!("Process PoX affirmations for reward cycle {}", reward_cycle);
+
+    let tx = burnchain_db.tx_begin()?;
+
+    let (prepare_ops, pox_anchor_block_info_opt) =
+        find_pox_anchor_block(&tx, reward_cycle, indexer, burnchain)?;
+
+    if let Some((anchor_block, descendancy)) = pox_anchor_block_info_opt.clone() {
+        debug!(
+            "PoX anchor block elected in reward cycle {} for reward cycle {} is {}",
+            reward_cycle,
+            reward_cycle + 1,
+            &anchor_block.block_header_hash
+        );
+
+        // anchor block found for this upcoming reward cycle
+        tx.set_anchor_block(&anchor_block, reward_cycle + 1)?;
+        assert_eq!(descendancy.len(), prepare_ops.len());
+
+        // mark the prepare-phase commits that elected this next reward cycle's anchor block as
+        // having descended or not descended from this anchor block.
+        for (block_idx, block_ops) in prepare_ops.iter().enumerate() {
+            assert_eq!(block_ops.len(), descendancy[block_idx].len());
+
+            for (tx_idx, tx_op) in block_ops.iter().enumerate() {
+                test_debug!(
+                    "Make affirmation map for block-commit at {},{}",
+                    tx_op.block_height,
+                    tx_op.vtxindex
+                );
+                tx.make_prepare_phase_affirmation_map(
+                    indexer,
+                    burnchain,
+                    reward_cycle + 1,
+                    tx_op,
+                    Some(&anchor_block),
+                    descendancy[block_idx][tx_idx],
+                )?;
+            }
+        }
+    } else {
+        debug!("PoX anchor block selected in reward cycle {} is None. Reward cycle {} has no anchor block", reward_cycle, reward_cycle + 1);
+
+        // anchor block not found for this upcoming reward cycle
+        tx.clear_anchor_block(reward_cycle + 1)?;
+
+        // mark the prepare-phase commits that did NOT elect this next reward cycle's anchor
+        // block as NOT having descended from any anchor block (since one was not chosen)
+        for block_ops in prepare_ops.iter() {
+            for tx_op in block_ops.iter() {
+                test_debug!(
+                    "Make affirmation map for block-commit at {},{}",
+                    tx_op.block_height,
+                    tx_op.vtxindex
+                );
+                tx.make_prepare_phase_affirmation_map(
+                    indexer,
+                    burnchain,
+                    reward_cycle + 1,
+                    tx_op,
+                    None,
+                    false,
+                )?;
+            }
+        }
+    }
+
+    tx.commit()?;
+    debug!(
+        "Processed PoX affirmations for reward cycle {}",
+        reward_cycle
+    );
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod test {
```

Comment:
I know we do this elsewhere, but I'd strongly favor separating _very large_ test modules out from the defining module -- some separation makes finding things and reviewing things in code much easier -- especially when > 2/3rds of this file is test cases.

Reply:
Wait, is it really 2/3 of the file? :scream:

Reply:
Refactored